docs/design-documents/20240430-schema-support.md

```diff
@@ -0,0 +1,413 @@
+make building schemas easier.
+
+### Schema service interface
+
+This section discusses the schema service's interface. Below we discuss options
+for the communication between Conduit and the connectors.
+
+#### Option 1: Stream of commands and responses
+
+This pattern is used in WASM processors. A server (in this case: Conduit)
+listens to commands (in this case: via a bidirectional gRPC stream). A client (
+in this case: a connector) sends a command to either register a schema or fetch
+a schema. Conduit receives the command and replies. An example can be seen below:
+
+```protobuf
+rpc CommandStream(stream Command) returns (stream Response);
+```
+
+For different types of commands and response to be supported, `Command`
+and `Response` need to have a `oneof` field where all the possible commands
+and respective responses are listed:
+
+```protobuf
+message Command {
+    oneof cmd {
+        SaveSchemaCommand saveSchemaCmd = 1;
+        // etc.
+   }
+}
+
+message Response {
+  oneof resp {
+    SaveSchemaResponse saveSchemaResp = 1;
+    // etc.
+  }
+}
+```    
+**Advantages**:
+
+1. No additional connection setup. When Conduit starts a connector process, it
+   establishes a connection. The same connection is used for all communication (
+   e.g. configuring a connector, opening, reading/writing records, etc.)
+2. Connector actions (which are planned for a future milestone) might use the
+   same command-and-reply stream.
+
+**Disadvantages**:
+
+1. A single method for all the operations makes both, the server and client
+   implementation, more complex. In Conduit, a single gRPC method needs to check
+   for the command type and then reply with a response. Then the client (i.e.
+   the connector) needs to check the response type. In case multiple commands
+   are sent, we need ordering guarantees.
+
+#### Option 2: Exposing a gRPC service in Conduit
+
+Conduit exposes a service to work with schemas. Connectors access the service
+and call methods on the service. 
+
+For this work, a connector (i.e. clients of the schema service) needs Conduit's
+IP address and the gRPC port. It's safe to assume that in most, if not all, real
+world use cases, Conduit and connectors will be running on the same host, so we
+can assume that the host is `localhost`. The gRPC port can be communicated to
+the connector via an environment variable.
+
+This service is intended to be used by connectors only. To facilitate that,
+Conduit can generate tokens that connectors would use to authenticate with
+Conduit. 
+
+The service should run on a random port. In VMs with hardened security, that
+might not be always possible, so it should be possible for the schema service to
+run on a pre-defined port.
+
+A skeleton of the gRPC definition of the service would be:
+
+```protobuf
+syntax = "proto3";
+
+service SchemaService {
+  rpc Create(CreateSchemaRequest) returns (CreateSchemaResponse);
+  rpc Fetch(FetchSchemaRequest) returns (FetchSchemaResponse);
+}
```

Comment:
I'm thinking if we should name this `ConnectorUtilService` and in the future extend it with more calls for other functionalities (e.g. storage).

```suggestion
service ConnectorUtilService {
  rpc CreateSchema(CreateSchemaRequest) returns (CreateSchemaResponse);
  rpc FetchSchema(FetchSchemaRequest) returns (FetchSchemaResponse);
  
  // Down the road:
  rpc Set(SetRequest) returns (SetResponse);
  rpc Get(GetRequest) returns (GetResponse);
  rpc Delete(DeleteRequest) returns (DeleteResponse);
}
```

We could as well leave this as `SchemaService` and then introduce `StorageService`. Maybe that's even better and we should leave it as it is. I'm just thinking out loud here 🙃

Reply:
I was thinking about something similar too... Naming it `ConnectorUtilService` would make it clear what clients should connect to it. On the other hand, having very different methods in a single service might make the service confusing. 

To me, the second option is slightly better. We can make it clear in the docs that connectors are supposed to be using the service (and it'll be running on a separate port too, so it'll make it even more clear). I don't think we'll have problems changing the name (since client code won't care much about it).

Reply:
Slightly unrelated, but I see that sometimes we refer as `create` and other times as `register` when it comes to the action of `creating` a schema. We should be consistent with this.